ykrt/src/compile/jitc_yk/codegen/x86_64.rs

```diff
@@ -0,0 +1,252 @@
+        // trace also get a 16-byte aligned stack, as per the SysV ABI.
+        self.align_asp(SYSV_CALL_STACK_ALIGN);
+
+        match i32::try_from(self.asp) {
+            Ok(asp) => {
+                let mut patchup = self.asm.alter_uncommitted();
+                patchup.goto(asm_off);
+                dynasm!(patchup
+                    // The size of this instruction must be the exactly the same as the dummy
+                    // allocation instruction that was emitted during `emit_prologue()`.
+                    ; sub rsp, DWORD asp
+                );
+            }
+            Err(_) => {
+                // If we get here, then the frame was so big that the dummy instruction we had
+                // planned to patch isn't big enough to encode the desired allocation size. Cross
+                // this bridge if/when we get to it.
+                todo!();
+            }
+        }
+    }
+
+    /// Load a local variable out of its stack slot into the specified register.
+    fn load_local(&mut self, reg: Rq, local: InstrIdx) {
+        let foff = self.allocation(local).frame_off();
+
+        match i32::try_from(foff) {
+            Ok(foff) => {
+                // We use `movzx` where possible to avoid partial register stalls.
+                match local.instr(self.jit_mod).def_abi_size() {
+                    1 => dynasm!(self.asm; movzx Rq(reg.code()), BYTE [rbp - foff]),
+                    2 => dynasm!(self.asm; movzx Rq(reg.code()), WORD [rbp - foff]),
+                    4 => dynasm!(self.asm; mov Rd(reg.code()), [rbp - foff]),
+                    8 => dynasm!(self.asm; mov Rq(reg.code()), [rbp - foff]),
+                    _ => todo!(),
+                }
+            }
+            Err(_) => todo!(),
+        }
+    }
+
+    /// Allocates storage space for the value a new local variable in the JITted code frame.
+    ///
+    /// This is platform specific because the frame offset depends on if the stack grows up or down.
+    fn allocate_local(&mut self, local: InstrIdx) -> LocalAlloc {
+        eprintln!("allocate: {:?}", local);
+        // Under the current design, there can't be gaps in `self.local_allocs` and local variable
+        // allocations happen sequentially. So the local we are currently allocating should be the
+        // next unallocated index.
+        debug_assert!(InstrIdx::new(self.local_allocs.len()).unwrap() == local);
+
+        // Align the stack to the size of the allocation.
+        let size = local.instr(self.jit_mod).def_abi_size();
+        self.align_asp(size);
+
+        // Allocate.
+        self.asp += size;
+        let alloc = LocalAlloc::new(self.asp);
+        self.local_allocs.push(alloc);
+        alloc
+    }
+
+    fn store_local(&mut self, l: &LocalAlloc, reg: Rq, size: usize) {
+        debug_assert!(l.frame_off <= self.asp);
+
+        match i32::try_from(l.frame_off) {
+            Ok(off) => match size {
```

Comment:
I suggest:

```rust
Ok(8) => ...
Ok(4) => ...
Err(...)
```

is clearer.

Reply:
I agree. Not sure why I did it like that.

Reply:
Hang on, that doesn't make sense. We are not matching the innards of the `Result`, but the `size` argument of the function.

The size and offset are independent here.

Reply:
Ah, oops! Although you could do `match (i32::try_from(...), size)` if you wanted, but it might not be clearer than what you currently have.

Reply:
Yeah, I think i prefer to do it the unclever way this time :)